compiler/codegen/CCData.hpp

```diff
@@ -199,11 +210,17 @@ class CCData
       bool find_unsafe(const key_t key, index_t * const index = NULL) const;
 
       /**
-       * @brief Converts a size in units of bytes to a size in units of data_t.
+       * @brief This function implements the actual work behind both put() and reserve().
        *
-       * @param[In] sizeBytes A size in bytes.
-       * @return A size in units of data_t.
+       * @param[In] value Optional. A pointer to the value to put.
+       * @param[In] sizeBytes The size of the value pointed to.
+       * @param[In] alignmentBytes The alignment (in bytes) to align the value to.
+       * @param[In] key Optional. The key to map the resulting index to. Without a key the index is the only reference to the data. If the key is already mapped to an index the operation will return the index and true, but no data will be written.
+       * @param[Out] index The index that refers to the value.
+       * @return True if the value exists in the table, or the key was already mapped to an index, false otherwise.
        */
+      bool put_impl(const void * const value, const size_t sizeBytes, const size_t alignmentBytes, const key_t * const key, index_t &index);
```

Comment:
Agreed, just tried the first suggestion since it sounds like less code overall, but I was saved by my own test cases.

It requires an extra `find()` for keyed values to determine if the key already exists, otherwise you risk overwriting an existing value. `reserve()` does a `find()` and returns `true` if the key exists, but `put()` can't rely on just that return value to gate the copy. The current implementation is the shorter and more optimal one.

The function is private already though, just hard to spot.